### PR TITLE
Implement SYCL as device backend

### PR DESCRIPTION
**Pull request by ZaubererHaft**
Merge the squashed changes implementing hipSYCL and oneAPI as device backend

### CODE REVIEW

File: DeviceMacros.h [hunk 1]
```diff
@@ -2,15 +2,18 @@
 #define DEVICE_MACROS_H
 
 #if defined(DEVICE_CUDA_LANG)
-#define DEVICE_KERNEL_LAUNCH(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, ...) \
-KRNL<<<GRID, BLOCK, SHR_SIZE, STREAM>>> ( __VA_ARGS__ )
+#define DEVICE_KERNEL_LAUNCH(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, ...)                                                 \
+  KRNL<<<GRID, BLOCK, SHR_SIZE, STREAM>>>(__VA_ARGS__)
```
Comment (ravil-mobile) [inline DeviceMacros.h]: too many white-spaces in line 5

File: DeviceMacros.h [hunk 2]
```diff
@@ -2,15 +2,18 @@
...
 
 #elif defined(DEVICE_HIP_LANG)
-#define DEVICE_KERNEL_LAUNCH(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, ...) \
-hipLaunchKernelGGL(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, __VA_ARGS__)
+#define DEVICE_KERNEL_LAUNCH(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, ...)                                                 \
+  hipLaunchKernelGGL(KRNL, GRID, BLOCK, SHR_SIZE, STREAM, __VA_ARGS__)
+
```
Comment (ravil-mobile) [inline DeviceMacros.h]: too many white-spaces in line 9

File: examples/basic/common.h [hunk 3]
```diff
@@ -6,11 +6,11 @@
 
 #if REAL_SIZE == 8
 using real = double;
-#define GEMM(LAYOUT, TRANSA, TRANSB, M, N, K, ALPHA, A, LDA, B, LDB, BETA, C, LDC)                 \
+#define GEMM(LAYOUT, TRANSA, TRANSB, M, N, K, ALPHA, A, LDA, B, LDB, BETA, C, LDC)                                     \
```
Comment (ravil-mobile) [inline examples/basic/common.h]: the same as before

File: examples/basic/common.h [hunk 4]
```diff
@@ -6,11 +6,11 @@
...
   cblas_dgemm(LAYOUT, TRANSA, TRANSB, M, N, K, ALPHA, A, LDA, B, LDB, BETA, C, LDC)
 #elif REAL_SIZE == 4
 using real = float;
-#define GEMM(LAYOUT, TRANSA, TRANSB, M, N, K, ALPHA, A, LDA, B, LDB, BETA, C, LDC)                 \
+#define GEMM(LAYOUT, TRANSA, TRANSB, M, N, K, ALPHA, A, LDA, B, LDB, BETA, C, LDC)                                     \
```
Comment (ravil-mobile) [inline examples/basic/common.h]: the same as before

File: examples/jacobi/CMakeLists.txt [hunk 5]
```diff
@@ -1,7 +1,8 @@
 cmake_minimum_required(VERSION 3.5)
+
 project(jsolver)
-set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
+#set(CMAKE_CXX_STANDARD 17)
+#set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (ravil-mobile) [inline examples/jacobi/CMakeLists.txt]: should be removed

File: examples/jacobi/CMakeLists.txt [hunk 6]
```diff
@@ -16,29 +17,37 @@ set(REAL_SIZE_IN_BYTES "8" CACHE STRING "size of the floating point data type")
 set_property(CACHE REAL_SIZE_IN_BYTES PROPERTY STRINGS "8" "4")
 
 set(DEVICE_BACKEND "CUDA" CACHE STRING "type of an interface")
-set_property(CACHE DEVICE_BACKEND PROPERTY STRINGS "CUDA" "HIP")
+set_property(CACHE DEVICE_BACKEND PROPERTY STRINGS "CUDA" "HIP" "ONEAPI" "HIPSYCL")
 
 set(SM "sm_60" CACHE STRING "model of Streaming Multiprocessor")
 set(SM_OPTIONS "sm_20" "sm_30" "sm_50" "sm_60" "sm_70" "gfx906")
 set_property(CACHE SM PROPERTY STRINGS ${SM_OPTIONS})
```
Comment (ravil-mobile) [inline examples/jacobi/CMakeLists.txt]: where are options for intel?

File: examples/jacobi/src/gpu/solver.cpp [hunk 7]
```diff
@@ -78,7 +79,7 @@ void gpu::Solver::run(const SolverSettingsT &settings, const CpuMatrixDataT &mat
         stream << "; comm: " << commStat.getStatistics().mean << ' ' << Statistics::getUnits();
 #endif
       }
-      Logger(ws, 0) << stream;
+      Logger(ws, ws.rank) << stream;
```
Comment (ravil-mobile) [inline examples/jacobi/src/gpu/solver.cpp]: What was a reason for this change?

File: examples/jacobi/src/host/solver.cpp [hunk 8]
```diff
@@ -72,10 +72,8 @@ void host::solver(const SolverSettingsT &settings, const CpuMatrixDataT &matrix,
           stream << "; comm: " << commStat.getStatistics().mean << ' ' << Statistics::getUnits();
 #endif
         }
-        Logger(ws, 0) << stream;
+        Logger(ws, ws.rank) << stream;
```
Comment (ravil-mobile) [inline examples/jacobi/src/host/solver.cpp]: again, why?

File: examples/jacobi/src/main.cpp [hunk 9]
```diff
@@ -22,7 +22,7 @@ int main(int argc, char *argv[]) {
     }
 
     YAML::Node params = YAML::LoadFile(argv[1]);
-    Logger(ws, 0) << (std::stringstream() << "Input parameters: \n" << params);
+    Logger(ws, ws.rank) << (std::stringstream() << "Input parameters: \n" << params);
```
Comment (ravil-mobile) [inline examples/jacobi/src/main.cpp]: the same as before

File: examples/jacobi/src/main.cpp [hunk 10]
```diff
@@ -39,12 +39,12 @@ int main(int argc, char *argv[]) {
   } catch (const std::runtime_error &error) {
     std::stringstream stream;
     stream << "Error: " << error.what();
-    Logger(ws, 0) << stream;
+    Logger(ws, ws.rank) << stream;
```
Comment (ravil-mobile) [inline examples/jacobi/src/main.cpp]: ?

File: examples/jacobi/src/main.cpp [hunk 11]
```diff
@@ -97,7 +97,7 @@ int main(int argc, char *argv[]) {
     stream << front.str() << "...\n" << middle.str() << "...\n" << back.str();
   }
 
-  Logger(ws, 0) << (stream << "Elapsed Time: " << time.count() << " [sec]");
+  Logger(ws, ws.rank) << (stream << "Elapsed Time: " << time.count() << " [sec]");
```
Comment (ravil-mobile) [inline examples/jacobi/src/main.cpp]: ?

File: examples/jacobi/src/main.cpp [hunk 12]
```diff
@@ -39,12 +39,12 @@ int main(int argc, char *argv[]) {
...
     std::stringstream stream;
     stream << "Error: " << error.what();
-    Logger(ws, 0) << stream;
+    Logger(ws, ws.rank) << stream;
     return -1;
   } catch (const std::logic_error &error) {
     std::stringstream stream;
     stream << "Error: " << error.what();
-    Logger(ws, 0) << stream;
+    Logger(ws, ws.rank) << stream;
```
Comment (ravil-mobile) [inline examples/jacobi/src/main.cpp]: ?

File: tests/reductions.cpp [hunk 13]
```diff
@@ -1,99 +1,277 @@
...
+
+  device->api->copy2dArrayTo(arr, dpitch, &hostVector[0], spitch, width, height);
+  int hostVector2[M][N + SPI];
+  device->api->copy2dArrayFrom(&hostVector2[0], spitch, arr, dpitch, width, height);
+
+  for (size_t i = 0; i < M; i++) {
+    for (size_t j = 0; j < N + SPI; j++) {
+      EXPECT_EQ(1904, hostVector2[i][j]);
+    }
+  }
```
Comment (ravil-mobile) [inline tests/reductions.cpp]: I like these tests. Well done! There is one little thing that should be fixed. I think these tests test Wrapped API and thus they should be moved to another test suite i.e. another file where each test starts from something like `TEST_F(API, ...)`

**Review by ravil-mobile — COMMENTED** [review]
Hi, everything looks good. Could you, please, move Wrapped API tests to a dedicate test suite?

File: DeviceMacros.h (hunk 1 above)
Comment (ZaubererHaft) [reply DeviceMacros.h]: whitespace is added automatically by clang-format, removed by hand. If we have time we could search for the belonging option

File: DeviceMacros.h (hunk 2 above)
Comment (ZaubererHaft) [reply DeviceMacros.h]: same as before

File: examples/basic/common.h (hunk 3 above)
Comment (ZaubererHaft) [reply examples/basic/common.h]: same as before

File: examples/basic/common.h (hunk 4 above)
Comment (ZaubererHaft) [reply examples/basic/common.h]: same as before

File: examples/jacobi/CMakeLists.txt (hunk 5 above)
Comment (ZaubererHaft) [reply examples/jacobi/CMakeLists.txt]: Removed

File: examples/jacobi/CMakeLists.txt (hunk 6 above)
Comment (ZaubererHaft) [reply examples/jacobi/CMakeLists.txt]: Add options for intel gpus received with `ocloc compile --help`

File: examples/jacobi/src/gpu/solver.cpp (hunk 7 above)
Comment (ZaubererHaft) [reply examples/jacobi/src/gpu/solver.cpp]: Logger should print MPI ranks properly running on different nodes (if not, every process prints "Info from rank 0" which does not help)

File: examples/jacobi/src/host/solver.cpp (hunk 8 above)
Comment (ZaubererHaft) [reply examples/jacobi/src/host/solver.cpp]: same as before

File: examples/jacobi/src/main.cpp (hunk 10 above)
Comment (ZaubererHaft) [reply examples/jacobi/src/main.cpp]: same as before

File: examples/jacobi/src/main.cpp (hunk 12 above)
Comment (ZaubererHaft) [reply examples/jacobi/src/main.cpp]: same as before

**Pull request merged**